torch/_library/define.py

```diff
@@ -0,0 +1,334 @@
+            >>>     assert weight.shape[0] == bias.shape[0]
+            >>>     assert x.device == weight.device
+            >>>     return x.new_empty(x.size(0), weight.size(0))
+            >>>
+            >>> x = torch.randn(2, 2)
+            >>> weight = torch.randn(2, 2)
+            >>> bias = torch.randn(2)
+            >>> # xdoctest: +SKIP("Requires Python <= 3.11")
+            >>> out = torch.compile(custom_linear, fullgraph=True)(x, weight, bias)
+            >>> assert torch.allclose(out, torch.nn.functional.linear(x, weight, bias))
+            >>>
+            >>> # Example 2: an operator with data-dependent output shape
+            >>> @torch.library.opaque_op(mutated_args=())
+            >>> def custom_nonzero(x: Tensor) -> Tensor:
+            >>>     x_np = x.cpu().numpy()
+            >>>     res = np.stack(np.nonzero(x_np), axis=1)
+            >>>     return torch.tensor(res, device=x.device)
+            >>>
+            >>> @custom_nonzero.impl_abstract
+            >>> def _(x):
+            >>>     # Number of nonzero-elements is data-dependent.
+            >>>     # Since we cannot peek at the data in an abstract impl,
+            >>>     # we use the ctx object to construct a new symint that
+            >>>     # represents the data-dependent size.
+            >>>     ctx = torch.library.get_ctx()
+            >>>     nnz = ctx.new_dynamic_size()
+            >>>     shape = [nnz, x.dim()]
+            >>>     result = x.new_empty(shape, dtype=torch.int64)
+            >>>     return result
+            >>>
+            >>> x = torch.tensor([0, 1, 2, 0, 0, 1])
+            >>> # xdoctest: +SKIP("Requires Python <= 3.11")
+            >>> out = torch.compile(custom_nonzero)(x)
+            >>> assert torch.allclose(out, x.nonzero())
+
+        """
+        self._abstract_fn = fn
+        return fn
+
+    def _build(self):
+        lib = self._lib
+        lib.define(f"{self._name}{self._schema}")
+        self._opoverload = _library.utils.lookup_op(self._qualname)
+
+        def abstract_impl(*args, **kwargs):
+            if self._abstract_fn is None:
+                import torch
+
+                raise torch._subclasses.fake_tensor.UnsupportedOperatorException(
+                    self._opoverload
+                )
+            return self._abstract_fn(*args, **kwargs)
+
+        library.impl_abstract(self._qualname, lib=lib)(abstract_impl)
+
+    def __call__(self, *args, **kwargs):
+        return self._opoverload(*args, **kwargs)
+
+
+OPDEF_TO_LIB: Dict[str, "library.Library"] = {}
+
+
+def get_library_allowing_overwrite(namespace: str, name: str) -> "library.Library":
+    qualname = f"{namespace}::{name}"
+
+    if qualname in OPDEF_TO_LIB:
+        OPDEF_TO_LIB[qualname]._destroy()
+        del OPDEF_TO_LIB[qualname]
+
+    lib = library.Library(namespace, "FRAGMENT")
+    OPDEF_TO_LIB[qualname] = lib
+    return lib
+
+
+def mangle_module(module):
+    """Mangles the module name.
```

Comment:
So... do you have an unmangler? Because that's the point of mangling

fwiw, I would suggest reserving __ as the dot separator, and then doing an escaping scheme for preexisting `__` in the name. Done this way, you don't need to have a `mangledN` signifier, the mangling is always the same.